### ISSUE TITLE
Excel says the document might have been saved by another person after some time

### ISSUE THREAD

**Issue by claell** [nextcloud/desktop#3456]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Expected behaviour
Excel documents should save without problems

### Actual behaviour
When having an Excel document open for some time and then trying to save it, a message appears saying that the file might have been modified by another person (which did not happen)

### Steps to reproduce
1. Use the Virtual Filesystem on Windows
2. Open an Excel file (save it)
3. Wait some time
4. Try to save it again

### Client configuration
Client version: 3.2.2

Operating system: Windows 10

OS language: German

Qt version used by client package (Linux only, see also Settings dialog):

Client package (From Nextcloud or distro) (Linux only):

Installation path of client:

### Server configuration

Nextcloud version:

Storage backend (external storage):

### Logs

Please use Gist (https://gist.github.com/) or a similar code paster for longer
logs.

1. Client logfile: 
Since 3.1: Under the "General" settings, you can click on "Create Debug Archive ..." to pick the location of where the desktop client will export the logs and the database to a zip file.
On previous releases: Via the command line: `nextcloud --logdebug --logwindow` or `nextcloud --logdebug --logfile log.txt`
(See also https://docs.nextcloud.com/desktop/3.0/troubleshooting.html#log-files)

2. Web server error log:

3. Server logfile: nextcloud log (data/nextcloud.log):

**Comment by claell** [comment]
Here is a German entry describing a similar problem with Excel and OneDrive: https://answers.microsoft.com/de-de/msoffice/forum/msoffice_excel-mso_win10-mso_365hp/excel-in-onedrive-durch-einen-anderen-benutzer/2fec7b7b-0114-4de5-99f3-952734afba4d?auth=1.

**Comment by claell** [comment]
This issue seems only to happen extremely rarely for me now. Maybe once or twice time since the last time I posted here.

**Comment by claell** [comment]
Still relevant. Just no dev has replied so far. It is not ideal that bugs are marked as stale in those cases.

**Comment by dahooz** [comment]
I have an observation that might help finding the reason for this issue.

For me, this problem occurs with version 3.3.2 on Windows for .tex files when using TeXStudio. The process is as follows:

1. I save a tex file from TeXStudio in some subfolder of the nextcloud sync folder, which is set to "virtual-mode".
2. after some seconds, Nextcloud Client notes the change and initiates sync (it must upload local to the server)
3. Immediately when I see the sync happening in nextcloud client settings window: TeXStudio complains, that there was an external change to the file.

I observed this situation in parallel with the Windows explorer: The reason might be a change on the *creation date* of the file. When TeXStudio saves the file, only the *modification date* and *access date* gets updated. As soon as Nextcloud Client syncs, the *creation date* is updated, too (to the same time as the *modification date* and *access date*). This seems to trigger the warning about an external change.

I have no idea of the WinAPI, but maybe you can not only [set the mod time (example code from downloading a file, where you update mod time)](https://github.com/nextcloud/desktop/blob/master/src/libsync/propagatedownload.cpp#L1060), but also backup and set the creation time?

Another reason could be a logic error: Could it be, that a file change (maybe a move or sth.) is triggered even if local is the most recent change? I can not see anything in [`propagateupload.cpp`](https://github.com/nextcloud/desktop/blob/master/src/libsync/propagateupload.cpp) (also not in `*ng.cpp` or `v1.cpp`) that could explain this. But I could imagine that despite local being the "newest version", still somehow something gets active that changes creation date, eg. Nextcloud Client downloads the file again, or triggers some slot related to downlaod?

**Comment by claell** [comment]
@dahooz Thanks for your observations! I am not sure, whether it explains the Excel problem, though (as the problem does not happen every time, in fact rather rarely, currently).

Still, the behavior of changing the creation date seems to be pretty wrong. Possibly this might even be worth a separate bug report to raise awareness.

Ping @allexzander @FlexW

**Comment by FlexW** [comment]
@dahooz Thanks for the valuable information. Maybe you can open an extra ticket for it when you have clear steps to reproduce the problem.

**Comment by claell** [comment]
@FlexW I know you are not working for Nextcloud anymore, but I think there are already clear steps for reproduction in the comment from @dahooz.

Apparently, this problem also happens for other files, according to @lars-sh. So maybe somebody can take a closer look, as this might explain some strange behavior of the sync and should also not happen.

**Comment by donaldsycamore** [comment]
Is this related to what I'm seeing with mtimes being truncated: can be reproduced at the bash/msys commandline (Windows 11, Nextcloud Desktop 3.4.1):

QOwnNotes is affected by this too.  I see that Nextcloud is truncating file timestamps -- running this in msys on Windows 11 I see after a few seconds the mtime changes:

```
donald@GIGABYTE-AERO MSYS /c/users/donald/nextcloud/test
$ while [ 1 ] ; do date |tee testfile; cat testfile; for i in `seq 1 10` ; do ls -l --time-style=full-iso testfile;sleep 1;done; done
Wed, Jan 26, 2022 10:47:27 PM
Wed, Jan 26, 2022 10:47:27 PM
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.491779500 +1300 testfile
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.491779500 +1300 testfile
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.491779500 +1300 testfile
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.491779500 +1300 testfile
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.000000000 +1300 testfile
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.000000000 +1300 testfile
-rw-r--r-- 1 donald donald 30 2022-01-26 22:47:27.000000000 +1300 testfile
```

**Comment by claell** [comment]
@donaldsycamore Thanks for the hint! This might indeed cause it, but I am not sure. Let's wait what the devs say.

@allexzander @mgallien can you reproduce this? Is this behavior wanted?

**Comment by donaldsycamore** [comment]
It looks like

* If you modify a file in an old-style sync folder, locally the mtime is not truncated, but if you sync it on another machine the mtime will be truncated
* If you modify a file in a virtual files sync folder, for ~5s the mtime is not trucated and then it is

So it's not just virtual files that is truncating the mtime, it's just that without virtual files that truncation is not visible on the machine where the change was made.

The file on the server has its mtime truncated too, so this is happening as the file is written to the server.  Is the sync protocol capable of sending the fraction-of-a-second mtime?

**Comment by claell** [comment]
Also note the comment above: https://github.com/nextcloud/desktop/issues/3456#issuecomment-914701354

There it is about creation times. Have you checked these as well?

**Comment by dahooz** [comment]
I think @donaldsycamore 's analysis leads somewhere: 

https://github.com/nextcloud/desktop/blob/fe799e983bf22f64d20ae4625ae0d77f7b5aa06e/src/libsync/filesystem.cpp#L73-L85

here the u_seconfs for both modification time and creation time get set to zero. This function gets called (only?) for the Virtual file system synchronization here:

https://github.com/nextcloud/desktop/blob/5e651878ff02514aa802e936d49b1b5626996ff7/src/libsync/vfs/xattr/vfs_xattr.cpp#L70-L78

and here

https://github.com/nextcloud/desktop/blob/5e651878ff02514aa802e936d49b1b5626996ff7/src/libsync/vfs/suffix/vfs_suffix.cpp#L69-L77

Which would explain why this is a problem only for the virtual file system.

the c_utimes is defined [here](https://github.com/nextcloud/desktop/blob/4efe00fecaf88b587c6bcb75f1af17ba1fa12747/src/csync/std/c_time.cpp), and I do not see a reason why useconds truncation is necessary.

Hunting down some git blames on [the line that truncates useconds](https://github.com/nextcloud/desktop/blob/fe799e983bf22f64d20ae4625ae0d77f7b5aa06e/src/libsync/filesystem.cpp#L77) explains nothing, unfortunately. The usecond truncation was inserted some time ago with [this commit](https://github.com/nextcloud/desktop/commit/7f8eba370045c12e4df8065daada66b6ea3cdf6e)

**Comment by donaldsycamore** [comment]
I get the feeling that Nextcloud/its predecessors simply don't expect to preserve subsecond mtimes.  Most of the files in my data/(user)/files folder on my server have truncated mtimes.  The oc_filecache table represents the mtime as a bigint.  Filesystem::setModTime() there is getting the modification time as a time_t, i.e. integer.

The options for fixing this then are either to make all of Nextcloud care about subsecond mtimes, or for the client to have a workaround where when resynchronizing an unmodified file it retains the tv_usec part of the mtime.

Such a workaround could be:

```
bool FileSystem::setModTime(const QString &filename, time_t modTime)
{
    struct timeval times[2];

    /* Avoid setting tv_usec to 0 if modification time has not changed */
    if (modTime == getModTime(filename)) {
        return true;
    }

    times[0].tv_sec = times[1].tv_sec = modTime;
    times[0].tv_usec = times[1].tv_usec = 0;
    int rc = c_utimes(filename, times);
    if (rc != 0) {
        qCWarning(lcFileSystem) << "Error setting mtime for" << filename
                                << "failed: rc" << rc << ", errno:" << errno;
        return false;
    }
    return true;
}
```

**Comment by mgallien** [comment]
> I get the feeling that Nextcloud/its predecessors simply don't expect to preserve subsecond mtimes. Most of the files in my data/(user)/files folder on my server have truncated mtimes. The oc_filecache table represents the mtime as a bigint. Filesystem::setModTime() there is getting the modification time as a time_t, i.e. integer.
> 
> The options for fixing this then are either to make all of Nextcloud care about subsecond mtimes, or for the client to have a workaround where when resynchronizing an unmodified file it retains the tv_usec part of the mtime.
> 
> Such a workaround could be [donaldsycamore@a295b0c](https://github.com/donaldsycamore/desktop/commit/a295b0c239c7f2627edc4d81647b2774fc62aedb)
> 
> I'm having trouble getting builds working locally so if someone could try out the patch and confirm it works (use testing method in [#3456 (comment)](https://github.com/nextcloud/desktop/issues/3456#issuecomment-1022035018)) that would be wonderful.

Thanks for your work on this !

you can open a pull request
as part of the automated checks pipeline, we trigger a build for Linux and for Windows
you would then get automated tests executed and we should at least see if nothing covered by them gets broken.

**Comment by claell** [comment]
An important thing to clarify is whether this is wanted or not. Of course, a workaround is nice, but it would be also nice to know whether the workaround is the actual solution or whether the actual solution is to make Nextcloud care about subsecond mtimes.

Pinging @tobiasKaminsky and also some people from the server side @nickvergessen @CarlSchwan @szaimen @skjnldsv

**Comment by WhiteChairFromIkea** [comment]
This issue should be renamed to "File is changed / touched after save in VFS (virtual file system)" or something like that, because this bug applies to any file, no excel relevant (or should I open another one?). Here are my steps to reproduce:

1. Have an txt file anywhere in VFS folder;
2. Open, edit, save it with text editor, capable of file change monitoring in the background ([notepad++](https://notepad-plus-plus.org/downloads/v8.2.1/), [notepad3](https://www.rizonesoft.com/downloads/));
3. Wait a bit. File just saved gets "touched" by nextcloud in some way (file attributes not changed, modification dates also, seems normal..);
4. Inside notepad: Get a notification that file is changed (aka "Do you want to reload file from disk?");

See the screencast: my file explorer ([DC commander](https://doublecmd.sourceforge.io/) highlights files which are "touched" in some way. See how txt file is touched unnecessary 2nd time after save:     

https://user-images.githubusercontent.com/46539653/162402085-273293ad-1449-4f4a-a91c-de35fe9e0a95.mp4

Expected behavior: File properties and content is not modified by nextcloud itself in any way. Follow the same steps to reproduce on Google VFS -> the issue is not present.

Client version:     
![3932a](https://user-images.githubusercontent.com/46539653/162415764-223a3490-2f3a-404b-b6f9-67900119efad.png)
(I am stuck at this version, because https://github.com/nextcloud/server/issues/3056#issuecomment-1092619556 and https://github.com/nextcloud/server/issues/30263 are not solved still (cannot risk to lose files in any way).)

**Comment by dsahlberg-apache-org** [comment]
There seems to be a pull request https://github.com/nextcloud/desktop/pull/4236

Is there anything I can do to help push this forward?

**Comment by bbceg** [comment]
Just to chip in (and apologies if this has already been made clear and I've missed it) but I'm seeing this without virtual folders. Perhaps the proposed fix doesn't care...

**Comment by tomdereub** [comment]
Same issue for me with excel 2021 on windows 10 pro, nextcloud desktop client 3.9.0 (without VFS) and Nextcloud 25.0.6.

![image](https://github.com/nextcloud/desktop/assets/16243499/c713420e-d85b-43df-b141-1dfba7eb3163)

Steps to reproduce exactly like said in the first post.

**Comment by roberix** [comment]
Hi. 

Same issue with client 3.9.0 with VSF on NC 25.0.8 using Windows 10 and Office 2019. 

Regards

rob

**Comment by bbceg** [comment]
I don't want to jump the gun but I think this may have been fixed in 3.9.2 (though I don't use Virtual Folders so perhaps it has only resolved the issue for local folders). 

Anyone else care to verify in case we can close this one?

**Comment by roberix** [comment]
Hi. 

Yes. We have been plagued by this. Gone since 3.9.2. 

Regards. 

Rob

**Comment by mgallien** [comment]
I posted a test build in the linked PR, people could you please test it and report your results here ?
that would be greatly appreciated and useful

**Comment by Mika-6-9** [comment]
Hi.
Same issue with client 3.10.2 with VSF on NC 27.1.1 using Windows 11.
It's not closed by #6212
Regards

**Comment by Mika-6-9** [comment]
Hi.
Same issue with client 3.11.0 with VSF on NC 27.1.1 using Windows 11.
It's not closed by https://github.com/nextcloud/desktop/pull/6263

I don't understand why it doesn't work, normally the change is included since version 3.10.2
Regards
